### PR TITLE
Write Output of AMW Run Script to Logfile

### PR DESCRIPTION
**Pull request by phil-pona**
Implemented first Version. @yvespp could you please review and Test on different AS and deployments?

### CODE REVIEW

File: AMW_business/src/main/java/ch/puzzle/itc/mobiliar/business/generator/control/RunSystemCallService.java [hunk 1]
```diff
@@ -50,6 +50,8 @@
 public class RunSystemCallService
 {
 
+	private String lineSeparator = System.getProperty("line.separator");
```
Comment (yvespp) [inline AMW_business/src/main/java/ch/puzzle/itc/mobiliar/business/generator/control/RunSystemCallService.java]: System.lineSeparator()

File: AMW_business/src/main/java/ch/puzzle/itc/mobiliar/business/generator/control/RunSystemCallService.java [hunk 2]
```diff
@@ -136,6 +147,7 @@ private void runSystemCall(String command)
 			InputStreamReader inread = new InputStreamReader(buf);
 			BufferedReader bufferedreader = new BufferedReader(inread);
 			StringBuilder sb = new StringBuilder();
+			StringBuilder systemCallResult = new StringBuilder();
 			String line;
 			while ((line = bufferedreader.readLine()) != null) {
 				// TODO schreibe resultat in eine logdatei
```
Comment (yvespp) [inline AMW_business/src/main/java/ch/puzzle/itc/mobiliar/business/generator/control/RunSystemCallService.java]: The TODOs can be removed now

**Comment by phil-pona** [comment]
very good point with the result as String, instead of passing a stream, I pass the path to the logfile. The interface seams smaller to me that way. I'm going to test the new integration first on my side.

**Comment by phil-pona** [comment]
@yvespp please check again and test on deployments

**Comment by yvespp** [comment]
Here is what I found in my tests:
* I removed all manual redirects in the rundscript like this:
  * old: `echo "start script" >> ${deployment.amwLogFile}`
  * new: `echo "start script"`
* The output shows up in the Log View, however the log of the predeploy disappears as soon as the deploy step starts. I guess this is because the file gets overwritten when den deploy starts. Can you instead append to the logo file if it already exists?
* I put `echo "@{test}"` in the deploy script which should show up in the deployment details but it doesn't.

@phil-pona can you fix the two points?

**Comment by yvespp** [comment]
Also stuff that gets logged to stderr will not show up in the log.
To get the error stream you can call `p.getErrorStream()`.

In ProcessBuilder there is an option to redirect stderr to stdout so you only have to deal with one stream: http://docs.oracle.com/javase/6/docs/api/java/lang/ProcessBuilder.html#redirectErrorStream(boolean)

**Comment by phil-pona** [comment]
- changed to append mode
- redirected ErrorStream to Logoutput
- values with @{test} are only displayed in the deployment view if an error occurs, this behavior did not change

**Comment by yvespp** [comment]
> values with @{test} are only displayed in the deployment view if an error occurs, this behavior did not change

Didn't know that, that's then.

@phil-pona I tested again, here is what I found:
* The Output gets redirected to a file with name `156704_node_01_script_output.log` but the ${deployment.amwLogFile} points to 156704_node_01.log. We write some output directly to the log using this variable which now ends up in the wrong file. This is needed because the predeploy script is run in the background. In the run script: `python -m refact.amwDeployRunner >> ${deployment.amwLogFile} 2>&1 &`
Can you make `deployment.amwLogFile` point to the actual log file?
* I don't see the annotated logs (e.g. `@{test}`) when the deployment fails. I tested this by adding an exit 1 at the end of the scripts. Maybe related to the first problem?
* Can you remove the brackets ouround this log: `Executing Script: (/data/amw/build/amw/Local_node_01_156707_predeploy/amwScript/AmwRunScript.sh)`
Looks a bit weird to me :-)

Such a small change, so much work 😃

**Comment by phil-pona** [comment]
fixed the two things, we need to test it well because of concurrency issues.

lets discuss the ${test} thing face to face next time

**Comment by phil-pona** [comment]
ok can be merged.

**Feature Breaking Change**
This pullrequest changes the behavior of the Deployment Logfile, since the java process now writes the script output directly into the same logfile. 
The output via ``${deployment.amwLogFile}`` in the deployment script appears first in the logfile.
The output created by the script with eg. (echo "test") will be logged after the other logoutput.

Workaround: choose a second Logfile to write asynch output to it. 
```
echo "test" >> ${deployment.targetLogPrefix}_secondlog.log
```

**Pull request merged**